In [47]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [48]:
#Define path
path = r'/Users/shikongo/Instacart Basket Analysis/'

In [49]:
path

'/Users/shikongo/Instacart Basket Analysis/'

In [50]:
df_prods = pd.read_csv(r'/Users/shikongo/Instacart Basket Analysis/02 Data/Original Data/Products.csv', index_col = False)

In [51]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Products.csv'), index_col = False)

In [52]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [53]:
# Check mixed data type
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [54]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [55]:
#Check missing values in the df_ prods 
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [56]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [57]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [58]:
# Address missing value
# with mean: df['column with missings'].fillna(mean value,inplace = True)
#with median: df'['column with missings'].fillna(median value, inplace = True)
df_prods_clean = df_prods[df_prods['product_name'].isnull()== False]

In [59]:
df_prods_clean.shape

(49677, 5)

In [60]:
#Find duplicate
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [61]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [62]:
# Address duplicate
df_prods_clean.shape

(49677, 5)

In [63]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [64]:
df_prods_clean_no_dups.shape

(49672, 5)

# df_ords

In [65]:
df_ords = pd.read_csv(r'/Users/shikongo/Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [66]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [67]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_wek  order_hour_of_day  days_since_prior_order  \
count       3.421083e+06       3.421083e+06            3.214874e+06   
mean        2.776219e+00       1.345202e+01            1.111484e+01   
std         2.046829e+00       4.226088e+00            9.206737e+00   
min         0.000000e+00       0.000000e+00            0.000000e+00   
25%         1.000000e+00       1.000000e+01            4.000000e+00   
50%         3.000000e+00       1.300000e+01            7.000000e+00   
75%         5.000000e+00       1.600000e+01            1.500000e+01   
max         6.000000e+00       2.300000e+01            3.000000e+01   

            ordr_id  
count  3.421083e+06  
mean   1.710542e+06  
std    9.875817e+05  
min    1.000000e+00  
25%    8.552715e+05  
50%    1.710542e+06  
75%    2.565812e+06  
max    3.421083e+06

# Interpret above output.

The describe() function provides a summary of statistics for each numerical column in the DataFrame. Here's an interpretation of the provided output and whether anything should be investigated further:

Count: All columns have the same count (3,421,083), indicating that there are no missing values in any of the columns.

Mean: The "mean" row shows the average value for each column. Most of the columns appear to have reasonable average values.

Std: The "std" row displays the standard deviation, which measures the dispersion of data. Standard deviations appear to be within reasonable ranges for most columns.

Min: The "min" row shows the minimum values for each column. Most minimum values are reasonable and within expected ranges. However, the "order_hour_of_day" and "orders_day_of_week" columns have a minimum value of 0, which could be unusual. Not sure if the data specifically codes these days and hours starting from 0.

25%: The 25th percentile indicates the value below which 25% of the data falls. These values appear to be reasonable.

50%: The 50th percentile (median) represents the middle value of the dataset. These values also seem reasonable.

75%: The 75th percentile indicates the value below which 75% of the data falls. These values are within reasonable ranges.

Max: The "max" row shows the maximum values for each column. Here, there are some columns with maximum values that are significantly higher than the 75th percentile. For example, the "order_number" column has a maximum value of 100, while the 75th percentile is 23. Similarly, the "user_id" and "ordr_id" columns have maximum values that are significantly higher than the 75th percentile. This could indicate potential outliers or data quality issues that may warrant further investigation.

In summary, while most of the statistics appear to be reasonable, it's essential to investigate the columns with maximum values that are significantly higher than the 75th percentile. These columns may contain outliers or data anomalies that need closer examination to ensure data quality and integrity.


In [68]:
# Check mixed -type data in the df_ords dataframes
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [69]:
df_ords['eval_set']= df_ords['eval_set'].astype('str')

In [70]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_wek  \
0           0   2539329        1    prior             1                  2   
1           1   2398795        1    prior             2                  3   
2           2    473747        1    prior             3                  3   
3           3   2254736        1    prior             4                  4   
4           4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  ordr_id  
0                  8                     NaN  2539329  
1                  7                    15.0  2398795  
2                 12                    21.0   473747  
3                  7                    29.0  2254736  
4                 15                    28.0   431534

Check mixed-type data and address it if necessary. Look at each column:

order_id: This column appears to contain integers, which is a single data type.

user_id: This column appears to contain integers, which is a single data type.

eval_set: This column appears to contain strings ("prior"), which is a single data type.

order_number: This column appears to contain integers, which is a single data type.

order_dow: This column appears to contain integers, which is a single data type.

order_hour_of_day: This column appears to contain integers, which is a single data type.

days_since_prior_order: This column contains both integers (e.g., 15) and a special value "NaN," which represents missing data.

It seems that the only column with mixed-type data is the "days_since_prior_order" column because it contains both integers and missing values represented by "NaN."

To fix this, I can convert the "days_since_prior_order" column to a consistent data type, such as float, to accommodate the missing values ("NaN". But I can fix this when I am going to address Mix data next.

In [71]:
# Check missing values in the df_ ords 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_wek              0
order_hour_of_day              0
days_since_prior_order    206209
ordr_id                        0
dtype: int64

In [72]:
df_prods.shape

(49693, 5)

In [73]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [74]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_wek  order_hour_of_day  days_since_prior_order  ordr_id  
0                        2                  8                     NaN  2539329  
11                       2                 11                     NaN  2168274  
26                       1                 14                     NaN  1374495  
39                       6                 11                     NaN  3343014  
45                       3                 12                     NaN  2717275  
...                    ...                ...                     ...      ...  
3420930                  4                 12                     NaN   969311  
3420934                  3                 18                     NaN  3189322  
3421002                  6                 19                     NaN  2166133  
3421019                  1                 15                     NaN  2227043  
3421069                  3                 11                     NaN  3154581  

[206209 rows x 9 columns]

# Report the findings and propose an explanation for any missing values find:

The "order_id," "user_id," "eval_set," "order_number," "order_dow," and "order_hour_of_day" columns have no missing values, as indicated by a count of 0 in the output.

Missing Values in "days_since_prior_order":

The "days_since_prior_order" column has 206,209 missing values, as indicated by the count of NaN values in the output.

Proposed Explanation:
The presence of missing values in the "days_since_prior_order" column may be attributed to the following reasons:

First-Time Orders: The missing values in this column may represent the first-time orders made by users. For users who are placing their first orders, there may not be a "days_since_prior_order" value because there is no prior order history to calculate the time since the previous order.

Data Collection Process: It's possible that during the data collection process, certain orders or user data were not captured accurately, leading to missing values in this column.

Data Entry Errors: Missing values could be the result of data entry errors or inconsistencies in the data source. Human errors or system glitches during data input can result in incomplete records.

User Behavior: Some users may not place orders at regular intervals, making it challenging to calculate the time since the prior order accurately. Users with irregular ordering patterns may have missing values in this column.




In [75]:
# Address the missing values using an appropriate method
df_ords['days_since_prior_order'].fillna(1.111484e+01,inplace = True)

In [76]:
df_ords.shape

(3421083, 9)

In [77]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [78]:
df_ords_clean.shape

(3421083, 9)

In [79]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_wek  \
0           0   2539329        1    prior             1                  2   
1           1   2398795        1    prior             2                  3   
2           2    473747        1    prior             3                  3   
3           3   2254736        1    prior             4                  4   
4           4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  ordr_id  
0                  8                11.11484  2539329  
1                  7                15.00000  2398795  
2                 12                21.00000   473747  
3                  7                29.00000  2254736  
4                 15                28.00000   431534

In [80]:
# Another method address missing value
df_ords.shape

(3421083, 9)

In [81]:
df_ords.dropna(inplace = True)

In [82]:
df_ords.shape

(3421083, 9)

# Explain why I used this method (filter) missing value

The method I used to address missing values in the "days_since_prior_order" column is to create a new DataFrame, df_ords_clean, by filtering out rows where the "days_since_prior_order" column is not null. This method is a valid approach to handle missing values, and it has certain advantages:

Choice Explanation:

Data Preservation: By using this method, I preserve all the rows with non-missing values in other columns, ensuring that no data points are lost in those columns. This can be important if I want to maintain the integrity of other information related to orders and users.

Simplicity: It's a straightforward and simple method to handle missing values. I am essentially excluding rows with missing "days_since_prior_order" values from my analysis.

No Imputation Assumption: This method avoids making any assumptions about the missing values. Instead of imputing values (e.g., using mean or median imputation), I am essentially saying that I want to use only the available data without making any guesses about the missing values.

Data Integrity: The resulting DataFrame, df_ords_clean, maintains the same structure as the original DataFrame, which can simplify subsequent analysis and maintain data integrity.



In [83]:
# Find the duplicates using an appropriate method
df_dups = df_ords_clean[df_ords_clean.duplicated()]


In [84]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_wek, order_hour_of_day, days_since_prior_order, ordr_id]
Index: []

# Check the duplicate values:

The output shows only the column names and no rows, it suggests that there are no rows that are exact duplicates of each other in the DataFrame when considering all columns collectively.

Propose an explanation method

Data Collection Process: If the data was collected systematically and accurately, it's less likely to contain duplicate entries. Data collection processes that prevent double counting or ensure data uniqueness can lead to a dataset without duplicates.

Database Design: If the dataset is sourced from a well-designed database, it may have constraints, such as primary keys or unique indexes, that prevent the insertion of duplicate records. Such database design practices help maintain data integrity.

Data Cleaning: Prior data cleaning and preprocessing efforts might have already identified and removed duplicate records from the dataset. Cleaning procedures often involve steps to eliminate duplicate entries to ensure data quality.

Unique Identifiers: Some columns in the dataset might serve as unique identifiers (e.g., "order_id" or "user_id"). If these columns are well-maintained and enforce uniqueness, they can prevent duplicates from being introduced.

Data Sources: Depending on the sources of the data, data providers or systems may have mechanisms in place to prevent or detect duplicate data before it gets into the dataset.

Data Sampling: If you're working with a representative sample of a larger dataset, it's possible that duplicates existed in the larger dataset but were not present in the sampled subset.

Data Entry Validation: Data entry processes with validation checks can catch and prevent duplicate data from being entered.



In [85]:
#Address duplicate value
#Since there is no duplicate, If was duplicate I can use this syntax to check the current number of rows
df_ords_clean.shape

(3421083, 9)

In [86]:
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

In [87]:
df_ords_clean_no_dups.shape

(3421083, 9)

In [88]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [89]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'prods_cleaned.csv'))

In [90]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_wek  \
0           0   2539329        1    prior             1                  2   
1           1   2398795        1    prior             2                  3   
2           2    473747        1    prior             3                  3   
3           3   2254736        1    prior             4                  4   
4           4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  ordr_id  
0                  8                11.11484  2539329  
1                  7                15.00000  2398795  
2                 12                21.00000   473747  
3                  7                29.00000  2254736  
4                 15                28.00000   431534

In [91]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3